agent/agent.go

```diff
@@ -237,6 +237,57 @@ func (a *Agent) GetRunningOutputPlugins() []string {
 	return res
 }
 
+type MapFieldSchema struct {
+	Value interface{}
+	Key   string
+}
+
+type ArrayFieldSchema struct {
+	Value  interface{}
+	Length int // Will be 0 if slice and array length if array
+}
+
+// GetPluginTypes returns a map of a plugin's field names to field value types
+func (a *Agent) GetPluginTypes(p interface{}) (map[string]interface{}, error) {
+
+	data := reflect.ValueOf(p).Elem() // extract Value of type interface{} from Value pointer to interface
+	schema := a.getFieldSchema(data.Type())
+
+	s, ok := schema.(map[string]interface{})
+
+	if ok {
+		return s, nil
+	}
+
+	return nil, fmt.Errorf("returned schema is not a map")
+}
+
+func (a *Agent) getFieldSchema(data reflect.Type) interface{} {
+	switch data.Kind() {
+	case reflect.Struct:
+		dataFields := make(map[string]interface{})
+		for i := 0; i < data.NumField(); i++ {
+			field := data.Field(i)
+			if field.PkgPath == "" { // only take exported fields
+				dataFields[field.Name] = a.getFieldSchema(field.Type)
+			}
+		}
+		return dataFields
+	case reflect.Array:
+		valueType := a.getFieldSchema(data.Elem())
+		return ArrayFieldSchema{valueType, data.Len()}
+	case reflect.Slice:
+		valueType := a.getFieldSchema(data.Elem())
+		return ArrayFieldSchema{valueType, 0}
```

Comment:
Why are slices and arrays being treated differently (A little confused about why the length is being set to 0)? A slice is essentially just a subarray of another array.

Reply:
We didn't think very hard about this but, we were concerned about a case that arrays might have a required length.

Reply:
Slices don't have a fixed length and arrays do right? We need some way to disambiguate between the two cases.